### PR TITLE
Simplify find_string operations

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

**Comment by nirvdrum** [comment]
I should note that this supplants #1045. If this is merged, I'll close out #1045.

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 1]
```diff
@@ -3479,24 +3478,31 @@ public Object stringToF(DynamicObject string, boolean strict,
...
     public static abstract class StringIndexPrimitiveNode extends CoreMethodArrayArgumentsNode {
 
-        @Child private StringByteCharacterIndexNode byteIndexToCharIndexNode = StringNodesFactory.StringByteCharacterIndexNodeFactory.create(null);
-        @Child private NormalizeIndexNode normalizeIndexNode = StringNodesFactory.NormalizeIndexNodeGen.create(null, null);
+        @Child private EncodingNodes.CheckEncodingNode checkEncodingNode;
+
+        @Specialization(guards = "start < 0")
+        public Object stringIndexBadStart(DynamicObject string, DynamicObject pattern, int start) {
+            throw new RaiseException(getContext().getCoreExceptions().argumentError("negative start given", this));
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: Does this happen in practice? Or can we just assert against it?

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 2]
```diff
@@ -3479,24 +3478,31 @@ public Object stringToF(DynamicObject string, boolean strict,
...
+
+        @Specialization(guards = { "start >= 0", "isEmpty(pattern)" })
+        public Object stringIndexEmptyPattern(DynamicObject string, DynamicObject pattern, int start) {
+            return start;
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: Does this happen in practice? Or can we just assert against it?

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 3]
```diff
@@ -3829,74 +3825,102 @@ public Object other(DynamicObject string, int index) {
...
+        @Specialization(guards = "start < 0")
+        public Object stringRindexBadStart(DynamicObject string, DynamicObject pattern, int start) {
+            throw new RaiseException(getContext().getCoreExceptions().argumentError("negative start given", this));
+        }
 
-            int pos = start;
+        @Specialization(guards = { "start >= 0", "isEmpty(pattern)" })
+        public Object stringRindexEmptyPattern(DynamicObject string, DynamicObject pattern, int start) {
+            return start;
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: Should this return the last index of the String?

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 4]
```diff
@@ -3829,74 +3825,102 @@ public Object other(DynamicObject string, int index) {
...
-        @Specialization(guards = "isRubyString(pattern)")
-        public Object stringRindex(DynamicObject string, DynamicObject pattern, int start,
-                @Cached("create()") BranchProfile errorProfile,
-                @Cached("create()") RopeNodes.BytesNode stringBytes,
-                @Cached("create()") RopeNodes.BytesNode patternBytes) {
-            // Taken from Rubinius's String::rindex.
+        @Specialization(guards = "start < 0")
+        public Object stringRindexBadStart(DynamicObject string, DynamicObject pattern, int start) {
+            throw new RaiseException(getContext().getCoreExceptions().argumentError("negative start given", this));
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: Does this happen in practice? Or can we just assert against it?

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 1 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: I think we can assert against this. I've looked at all the callers and it appears they try to deal with a negative value before every making the call.

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 2 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: The string can be empty in practice.

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: Let's `assert` then, it makes for simpler and more straightforward code.

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 3 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: Not according to the original Rubinius implementation. And specs are all passing with it this way (it's also the way it worked in the old implementation).

Comment (eregon) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: Weird. It doesn't make sense to me to return the first index if searching from the end.
Do the specs not pass if we return `string.length`?
I'd guess it's a Rubinius bug.

**Pull request merged**